sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,389 @@
+
+        floating *centers_new_chunk
+        floating *weight_in_clusters_chunk
+        floating *pairwise_distances_chunk
+
+    # count remainder chunk in total number of chunks
+    n_chunks += n_samples != n_chunks * n_samples_chunk
+
+    if update_centers:
+        memset(&centers_new[0, 0], 0, n_clusters * n_features * sizeof(floating))
+        memset(&weight_in_clusters[0], 0, n_clusters * sizeof(floating))
+
+    with nogil, parallel(num_threads=n_jobs):
+        # thread local buffers
+        centers_new_chunk = <floating*> calloc(n_clusters * n_features, sizeof(floating))
+        weight_in_clusters_chunk = <floating*> calloc(n_clusters, sizeof(floating))
+        pairwise_distances_chunk = <floating*> malloc(n_samples_chunk * n_clusters * sizeof(floating))
+
+        for chunk_idx in prange(n_chunks):
+            start = chunk_idx * n_samples_chunk
+            if chunk_idx == n_chunks - 1 and n_samples_rem > 0:
+                end = start + n_samples_rem
+            else:
+                end = start + n_samples_chunk
+
+            _update_chunk_dense(
+                &X[start, 0],
+                sample_weight[start: end],
+                x_squared_norms[start: end],
+                centers_old,
+                centers_squared_norms,
+                labels[start: end],
+                centers_new_chunk,
+                weight_in_clusters_chunk,
+                pairwise_distances_chunk,
+                update_centers)
+
+        # reduction from local buffers. The gil is necessary for that to avoid
+        # race conditions.
+        if update_centers:
+            with gil:
+                for j in range(n_clusters):
+                    weight_in_clusters[j] += weight_in_clusters_chunk[j]
+                    for k in range(n_features):
+                        centers_new[j, k] += centers_new_chunk[j * n_features + k]
+
+        free(centers_new_chunk)
+        free(weight_in_clusters_chunk)
+        free(pairwise_distances_chunk)
+
+    if update_centers:
+        _relocate_empty_clusters_dense(X, sample_weight, centers_old,
+                                    centers_new, weight_in_clusters, labels)
+
+        _average_centers(centers_new, weight_in_clusters)
+        _center_shift(centers_old, centers_new, center_shift)
+
+
+cdef void _update_chunk_dense(floating *X,
+                              floating[::1] sample_weight,
+                              floating[::1] x_squared_norms,
+                              floating[:, ::1] centers_old,
+                              floating[::1] centers_squared_norms,
+                              int[::1] labels,
+                              floating *centers_new,
+                              floating *weight_in_clusters,
+                              floating *pairwise_distances,
+                              bint update_centers) nogil:
+    """K-means combined EM step for one dense data chunk.
```

Comment:
Maybe this function is actually too hard to test because it would require writing a wrapper just for the test.

Instead we would have a test that uses the public API to run only a single K-Means iteration from manually initialized centroids and compare the results to a naive K-Means iteration written in pure numpy.

This unit test would make it easier to debug any future regression (compared to tests that run many iterations and compare the final result and inertia).

Reply:
done